### PR TITLE
[drake_cmake_external] Rebuild eigen,fmt,spdlog from source

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
Test coverage for RobotLocomotion/drake#16652 and https://github.com/RobotLocomotion/drake/pull/16700.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/218)

### CODE REVIEW

File: drake_cmake_external/CMakeLists.txt [hunk 1]
```diff
@@ -69,7 +69,52 @@ find_package(PythonInterp ${FIND_PYTHON_INTERP_VERSION} MODULE REQUIRED)
...
+    -DCMAKE_CXX_COMPILER:FILEPATH=${CMAKE_CXX_COMPILER}
+    -DCMAKE_INSTALL_PREFIX:PATH=${CMAKE_INSTALL_PREFIX}
+    -DCMAKE_PREFIX_PATH:PATH=${CMAKE_PREFIX_PATH}
+    -DCMAKE_VERBOSE_MAKEFILE:BOOL=${CMAKE_VERBOSE_MAKEFILE}
+    -DBUILD_SHARED_LIBS=ON
+    -DCMAKE_CXX_FLAGS:STRING=-fPIC
+    -DFMT_INC_DIR:PATH=include/fmt
+    -DFMT_TEST=OFF
+  BUILD_ALWAYS ON
+  INSTALL_COMMAND :
```
Comment (mwoehlke-kitware) [inline drake_cmake_external/CMakeLists.txt:92]: This looks suspicious. I am using "installed" (to siblings of my build directory, not anything like `/usr/local`) versions of both fmt and spdlog, as at least one of them was totally unusable from a build tree.

Also, spdlog almost certainly needs `fmt_DIR` specified.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/218)* status: 1 unresolved discussion, platform LGTM missing (waiting on @mwoehlke-kitware)

___
*[drake_cmake_external/CMakeLists.txt, line 92 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/218#-Mx0lAJ6nVBVQt2iMw9r-r1-92:-Mx0wfez-R-3op6f0Hf8:b-68c24m) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/e646ae7ffae4b90a53824b7714d36e2bf61e3252/drake_cmake_external/CMakeLists.txt#L92)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

This looks suspicious. I am using "installed" (to siblings of my build directory, not anything like `/usr/local`) versions of both fmt and spdlog, as at least one of them was totally unusable from a build tree.

Also, spdlog almost certainly needs `fmt_DIR` specified.
</blockquote></details>

Indeed, that's an inadvertent paste-o from the Drake rule below, that I was copying from.

Could you prepare a patch (for this, any any other bugs) that passes your own local testing?

File: drake_cmake_external/CMakeLists.txt (hunk 1 above)
Comment (mwoehlke-kitware) [inline drake_cmake_external/CMakeLists.txt:92]:
```suggestion
```

File: drake_cmake_external/CMakeLists.txt [hunk 2]
```diff
@@ -69,7 +69,52 @@ find_package(PythonInterp ${FIND_PYTHON_INTERP_VERSION} MODULE REQUIRED)
...
+  CMAKE_ARGS
+    -DCMAKE_BUILD_TYPE:STRING=${CMAKE_BUILD_TYPE}
+    -DCMAKE_C_COMPILER:FILEPATH=${CMAKE_C_COMPILER}
+    -DCMAKE_CXX_COMPILER:FILEPATH=${CMAKE_CXX_COMPILER}
+    -DCMAKE_INSTALL_PREFIX:PATH=${CMAKE_INSTALL_PREFIX}
+    -DCMAKE_PREFIX_PATH:PATH=${CMAKE_PREFIX_PATH}
+    -DCMAKE_VERBOSE_MAKEFILE:BOOL=${CMAKE_VERBOSE_MAKEFILE}
+    -DSPDLOG_BUILD_SHARED=ON
+    -DCMAKE_CXX_FLAGS:STRING=-fPIC
+    -DSPDLOG_FMT_EXTERNAL:BOOLEAN=ON
```
Comment (mwoehlke-kitware) [inline drake_cmake_external/CMakeLists.txt:113]:
```suggestion
    -DSPDLOG_FMT_EXTERNAL:BOOLEAN=ON
    -Dfmt_DIR=${CMAKE_INSTALL_PREFIX}/lib/cmake/fmt
 ```

**Comment by mwoehlke-kitware** [comment]
___
*[drake_cmake_external/CMakeLists.txt, line 92 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/218#-Mx0lAJ6nVBVQt2iMw9r-r1-92:-Mx0y6d-8HRqBImHOaMF:b9zrcmj) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/e646ae7ffae4b90a53824b7714d36e2bf61e3252/drake_cmake_external/CMakeLists.txt#L92)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

```suggestion
```
</blockquote></details>

How would I test? It seems to need some sort of local changes to pull robotlocomotion/drake#16652?

**Review by jwnimmer-tri — COMMENTED** [review]
> Dismissed @mwoehlke-kitware from a discussion.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/218)* status: 1 unresolved discussion, platform LGTM missing (waiting on @mwoehlke-kitware)

___
*[drake_cmake_external/CMakeLists.txt, line 92 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/218#-Mx0lAJ6nVBVQt2iMw9r-r1-92:-Mx0zjtACYYag-WNBl_w:bdk1duj) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/e646ae7ffae4b90a53824b7714d36e2bf61e3252/drake_cmake_external/CMakeLists.txt#L92)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

How would I test? It seems to need some sort of local changes to pull robotlocomotion/drake#16652?
</blockquote></details>

OK I hadn't be able to get far enough in the build to check that.  With your suggested fixes, I was able to reach the Drake build and point it at the Drake PR.

In my own local test, I got far enough here that Drake started building.  I think if CI is happy here, then that's mostly good enough.

The one thing I'd still ask you to check locally is that the build here is _in fact_ using the custom fmt and spdlog.  You can probably do that by poking at the files, rather than waiting for a full build?

**Comment by mwoehlke-kitware** [comment]
___
*[drake_cmake_external/CMakeLists.txt, line 92 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/218#-Mx0lAJ6nVBVQt2iMw9r-r1-92:-Mx1AKJN3yxlLtcGtwno:b-8bnu2d) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/e646ae7ffae4b90a53824b7714d36e2bf61e3252/drake_cmake_external/CMakeLists.txt#L92)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

OK I hadn't be able to get far enough in the build to check that.  With your suggested fixes, I was able to reach the Drake build and point it at the Drake PR.

In my own local test, I got far enough here that Drake started building.  I think if CI is happy here, then that's mostly good enough.

The one thing I'd still ask you to check locally is that the build here is _in fact_ using the custom fmt and spdlog.  You can probably do that by poking at the files, rather than waiting for a full build?
</blockquote></details>

I'm not sure what "poking at files" you had in mind?

It built for me, and I do see:
```
root@b358281bdc2f:~/drake-external-examples/build# ldd install/lib/libdrake.so | grep -e fmt -e spd
        libfmt.so.6 => /root/drake-external-examples/build/install/lib/libfmt.so.6 (0x00007f7f1c72d000)
        libspdlog.so.1 => /root/drake-external-examples/build/install/lib/libspdlog.so.1 (0x00007f7f1c6a3000)
```

On a "normal" drake Build, there is no libfmt, and libspdlog comes from `/usr/lib`. So it seems to be using the user-provided ones? (Note also this was built on a Docker Focal container that has no fmt/spdlog installed.)

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/218)* status: 1 unresolved discussion, platform LGTM missing (waiting on @jwnimmer-tri and @mwoehlke-kitware)

___
*[drake_cmake_external/CMakeLists.txt, line 92 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/218#-Mx0lAJ6nVBVQt2iMw9r-r1-92:-Mx1BrWV8RhJ8vupODWC:bd50uif) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/e646ae7ffae4b90a53824b7714d36e2bf61e3252/drake_cmake_external/CMakeLists.txt#L92)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

I'm not sure what "poking at files" you had in mind?

It built for me, and I do see:
```
root@b358281bdc2f:~/drake-external-examples/build# ldd install/lib/libdrake.so | grep -e fmt -e spd
        libfmt.so.6 => /root/drake-external-examples/build/install/lib/libfmt.so.6 (0x00007f7f1c72d000)
        libspdlog.so.1 => /root/drake-external-examples/build/install/lib/libspdlog.so.1 (0x00007f7f1c6a3000)
```

On a "normal" drake Build, there is no libfmt, and libspdlog comes from `/usr/lib`. So it seems to be using the user-provided ones? (Note also this was built on a Docker Focal container that has no fmt/spdlog installed.)
</blockquote></details>

Good so far.  I'll run this on my main Focal machine that _does_ have spdlog installed, and make sure I see the same.  That should be sufficient.

**Comment by jwnimmer-tri** [comment]
@mwoehlke-kitware could you also look at the Jenkins failure here?  Do we have a too-old CMake or something?

**Comment by mwoehlke-kitware** [comment]
Seems likely, based on the error. It looks like [`CREATE_LINK`](https://cmake.org/cmake/help/v3.14/command/file.html#create-link) was added in CMake 3.14.

**Review by jwnimmer-tri — COMMENTED** [review]
=> OK, filed the CI upgrade as #219.  We can hold off merging this PR until the CI swap is ready.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/218#-:-Mx1UJso-PaaBl8PDHV2:bpmzsxc)* status: 1 unresolved discussion, platform LGTM missing (waiting on @mwoehlke-kitware)

___
*[drake_cmake_external/CMakeLists.txt, line 92 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/218#-Mx0lAJ6nVBVQt2iMw9r-r1-92:-Mx1Uq6t2KM_V6UR6gIW:b-ske5un) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/e646ae7ffae4b90a53824b7714d36e2bf61e3252/drake_cmake_external/CMakeLists.txt#L92)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Good so far.  I'll run this on my main Focal machine that _does_ have spdlog installed, and make sure I see the same.  That should be sufficient.
</blockquote></details>

OK, this also looks great on a Focal build with Ubuntu's spdlog installed.

```
jwnimmer@call-cps:~/jwnimmer-tri/x-drake-external-examples/drake_cmake_external$ ldd build/install/lib/libdrake.so | grep -e fmt -e spd
	libvtkfmt-9.1.so.1 => /home/jwnimmer/jwnimmer-tri/x-drake-external-examples/drake_cmake_external/build/install/lib/libvtkfmt-9.1.so.1 (0x00007f2bcf9e4000)
	libfmt.so.6 => /home/jwnimmer/jwnimmer-tri/x-drake-external-examples/drake_cmake_external/build/install/lib/libfmt.so.6 (0x00007f2bcdf30000)
	libspdlog.so.1 => /home/jwnimmer/jwnimmer-tri/x-drake-external-examples/drake_cmake_external/build/install/lib/libspdlog.so.1 (0x00007f2bcdea6000)
```

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/218)* status: 1 unresolved discussion, platform LGTM missing (waiting on @mwoehlke-kitware)

___
*[drake_cmake_external/CMakeLists.txt, line 92 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/218#-Mx0lAJ6nVBVQt2iMw9r-r1-92:-Mxv91uK1AN_KpVoUsvC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/e646ae7ffae4b90a53824b7714d36e2bf61e3252/drake_cmake_external/CMakeLists.txt#L92)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

OK, this also looks great on a Focal build with Ubuntu's spdlog installed.

```
jwnimmer@call-cps:~/jwnimmer-tri/x-drake-external-examples/drake_cmake_external$ ldd build/install/lib/libdrake.so | grep -e fmt -e spd
	libvtkfmt-9.1.so.1 => /home/jwnimmer/jwnimmer-tri/x-drake-external-examples/drake_cmake_external/build/install/lib/libvtkfmt-9.1.so.1 (0x00007f2bcf9e4000)
	libfmt.so.6 => /home/jwnimmer/jwnimmer-tri/x-drake-external-examples/drake_cmake_external/build/install/lib/libfmt.so.6 (0x00007f2bcdf30000)
	libspdlog.so.1 => /home/jwnimmer/jwnimmer-tri/x-drake-external-examples/drake_cmake_external/build/install/lib/libspdlog.so.1 (0x00007f2bcdea6000)
```
</blockquote></details>

Done.

**Comment by jwnimmer-tri** [comment]
@mwoehlke-kitware ping -- waiting on your approval / LGTM.

**Comment by mwoehlke-kitware** [comment]
:lgtm: I suppose, though I have not attempted to build it myself. Do you need me to do that?

**Review by jwnimmer-tri — COMMENTED** [review]
> Dismissed @mwoehlke-kitware from a discussion.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/218)* status: all discussions resolved, platform LGTM missing (waiting on @mwoehlke-kitware)

**Comment by jwnimmer-tri** [comment]
I think the CI coverage should suffice on its own.

(Self-:lgtm: to placate reviewable.)

**Pull request merged**